rclc/src/rclc/action_server.c

```diff
@@ -0,0 +1,325 @@
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    ros_request, "request is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    request_header, "request_header is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+
+  if(!rclc_action_server_is_valid_handle(rclc_handle)){
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  Generic_GetResult_Request * request = (Generic_GetResult_Request *) ros_request;
+
+  bool goal_id_match = 0 == memcmp(request->goal_id.uuid, rclc_handle->goal_request->goal_id.uuid, sizeof(request->goal_id.uuid));
+
+  if (!goal_id_match) {
+    return RCL_RET_ERROR;
+  }
+
+  rclc_handle->get_result_request = request_header;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rclc_action_server_finish_goal_sucess(
+  rclc_action_goal_handle_t rclc_handle,
+  void * ros_response)
+{
+  if(!rclc_action_server_is_valid_handle(rclc_handle)){
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  Generic_GetResult_Response * response = (Generic_GetResult_Response *)ros_response;
+
+  rcl_ret_t rc = rcl_action_update_goal_state(rclc_handle->rcl_handle, GOAL_EVENT_SUCCEED);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_finish_goal_sucess, rcl_action_update_goal_state);
+    return rc;
+  }
+
+  rc = rcl_action_notify_goal_done(rclc_handle->action_server);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_finish_goal_sucess, rcl_action_notify_goal_done);
+    return rc;
+  }
+
+  response->status = action_msgs__msg__GoalStatus__STATUS_SUCCEEDED;
+
+  return rcl_action_send_result_response(rclc_handle->action_server, rclc_handle->get_result_request, response);
+}
+
+rcl_ret_t
+rclc_action_server_finish_goal_abort(
+  rclc_action_goal_handle_t rclc_handle)
+{
+  if(!rclc_action_server_is_valid_handle(rclc_handle)){
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  rcl_ret_t rc = rcl_action_update_goal_state(rclc_handle->rcl_handle, GOAL_EVENT_ABORT);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_finish_goal_abort, rcl_action_update_goal_state);
+    return rc;
+  }
+
+  rc = rcl_action_notify_goal_done(rclc_handle->action_server);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_finish_goal_abort, rcl_action_notify_goal_done);
+    return rc;
+  }
+
+  Generic_GetResult_Response response;
+  response.status = action_msgs__msg__GoalStatus__STATUS_ABORTED;
+
+  return rcl_action_send_result_response(rclc_handle->action_server, rclc_handle->get_result_request, &response);
+}
+
+rcl_ret_t
+rclc_action_server_finish_goal_cancel(
+  rclc_action_goal_handle_t rclc_handle,
+  action_msgs__srv__CancelGoal_Request * request,
+  rmw_request_id_t * request_header)
+{
+  if(!rclc_action_server_is_valid_handle(rclc_handle)){
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  bool goal_id_match = 0 ==
+    memcmp(request->goal_info.goal_id.uuid, rclc_handle->goal_request->goal_id.uuid, sizeof(request->goal_info.goal_id.uuid));
+
+  if (goal_id_match) {
```

Comment:
I propose to check for `!goal_id_match` and return immediately with `RCL_RET_OK` (cf. line 283) to reduce nesting and complexity.